scripts/wallet.js

```diff
@@ -672,15 +676,54 @@ export class Wallet {
                     type,
                     tx.txid,
                     arrReceivers,
-                    false,
+                    tx.hasShieldData,
                     tx.blockTime,
                     tx.blockHeight,
-                    Math.abs(nAmount)
+                    nAmount,
+                    nShieldAmount,
+                    ownAllVin && ownAllVout && ownAllShield
                 )
             );
         }
         return histTXs;
     }
+
+    /**
+     * Extract the sapling spent, received and shield addressed, regarding the wallet, from a tx
+     * @param {import('./transaction.js').Transaction} tx - a Transaction object
+     */
+    async extractSaplingAmounts(tx) {
+        let shieldCredit = 0;
+        let shieldDebit = 0;
+        let arrShieldReceivers = [];
+        if (!tx.hasShieldData || !wallet.hasShield()) {
+            return { shieldCredit, shieldDebit, arrShieldReceivers };
+        }
+
+        for (const shieldSpend of tx.shieldSpend) {
+            //TODO: find a better way to reverse and swap endianess
+            const stringArr = shieldSpend.nullifier.split('').reverse();
+            for (let i = 0; i < stringArr.length - 1; i += 2) {
+                const temp = stringArr[i + 1];
+                stringArr[i + 1] = stringArr[i];
+                stringArr[i] = temp;
+            }
```

Comment:
You can convert to array, swap that and re-convert to hex

Reply:
Imo, we can write an efficient `swapEndianness` util in `util.js`, useful and easier to read to re-use in a compact form than manually swap it for every instance.